lms/lib/courseware_search/lms_filter_generator.py

```diff
@@ -8,37 +8,88 @@
 from courseware.access import get_user_role
 
 
+INCLUDE_SCHEMES = [CohortPartitionScheme, RandomUserPartitionScheme, ]
+SCHEME_SUPPORTS_ASSIGNMENT = [RandomUserPartitionScheme, ]
+
+
 class LmsSearchFilterGenerator(SearchFilterGenerator):
     """ SearchFilterGenerator for LMS Search """
 
+    _user_enrollments = {}
+
+    def _enrollments_for_user(self, user):
+        """ Return users course enrollments """
+        if user not in self._user_enrollments:
+            self._user_enrollments[user] = CourseEnrollment.enrollments_for_user(user)
+        return self._user_enrollments[user]
+
     def filter_dictionary(self, **kwargs):
-        """ base implementation which filters via start_date """
-        filter_dictionary = super(LmsSearchFilterGenerator, self).filter_dictionary(**kwargs)
-        if 'user' in kwargs and 'course_id' in kwargs and kwargs['course_id']:
-            user = kwargs['user']
-            try:
-                course_key = CourseKey.from_string(kwargs['course_id'])
-            except InvalidKeyError:
-                course_key = SlashSeparatedCourseKey.from_deprecated_string(kwargs['course_id'])
-
-            # Staff user looking at course as staff user
-            if get_user_role(user, course_key) == 'staff':
-                return filter_dictionary
-
-            cohorted_user_partition = get_cohorted_user_partition(course_key)
-            if cohorted_user_partition:
-                partition_group = cohorted_user_partition.scheme.get_group_for_user(
+        """ LMS implementation, adds filtering by content groups, course id and user """
+
+        def get_group_for_user_partition(user_partition, course_key, user):
+            """ Returns users group for user partition """
+            if user_partition.scheme in SCHEME_SUPPORTS_ASSIGNMENT:
+                return user_partition.scheme.get_group_for_user(
+                    course_key,
+                    user,
+                    user_partition,
+                    assign=False,
+                )
+            else:
+                return user_partition.scheme.get_group_for_user(
                     course_key,
                     user,
-                    cohorted_user_partition,
+                    user_partition,
                 )
-                filter_dictionary['content_groups'] = unicode(partition_group.id) if partition_group else None
+
+        def get_content_groups(course, user):
+            """ Collect content groups for user for this course """
+            partition_groups = []
+            for user_partition in course.user_partitions:
+                if user_partition.scheme in INCLUDE_SCHEMES:
+                    group = get_group_for_user_partition(user_partition, course.id, user)
+                    if group:
+                        partition_groups.extend(group)
+            content_groups = [unicode(partition_group.id) for partition_group in partition_groups if partition_group]
```

Comment:
Original implementation wasn't working because in case group was None it would throw NoneType not iterable error. This happens in cases where group is not automatically assigned.
(I agree, if partition_group is not needed any more)